databricks_cli/tunnel/api.py

```diff
@@ -0,0 +1,242 @@
+        resp = self.command_client. \
+            execute_command_until_terminated(language="python",
+                                             cluster_id=self.cluster_id,
+                                             context_id=self.context_id,
+                                             command=install_keys_on_driver_cmd)
+        if resp["status"] != "Finished":
+            err_data = resp.get("results", {}).get("data", "")
+            raise RuntimeError(f"Sending public key to driver was not successful: {err_data}")
+        if resp["results"]["resultType"] == "error":
+            err_data = resp.get("results", {}).get("cause", "")
+            raise RuntimeError(f"Sending public key to driver was not successful: {err_data}")
+
+    @property
+    def default_remote_ssh_dir(self):
+        return self._default_remote_ssh_dir
+
+    @default_remote_ssh_dir.setter
+    def default_remote_ssh_dir(self, value):
+        self._default_remote_ssh_dir = value
+
+    @property
+    def default_local_ssh_dir(self):
+        return self._default_local_ssh_dir
+
+    @default_local_ssh_dir.setter
+    def default_local_ssh_dir(self, value):
+        self._default_local_ssh_dir = value
+
+    @property
+    def local_private_key_path(self):
+        return f"{self.default_local_ssh_dir}/{self.cluster_id}"
+
+    @property
+    def local_public_key_path(self):
+        return f"{self.local_private_key_path}.pub"
+
+    def setup_ssh_keys(self):
+        print("Setting up SSH keys...")
+
+        private_key, public_key = generate_key_pair()
+        self.send_public_key_to_driver(public_key)
+        self.setup_local_keys(private_key, public_key)
+
+    def setup_local_keys(self, private_key, public_key):
+        private_key_path = Path(self.local_private_key_path).expanduser()
+        private_key_path.parent.mkdir(parents=True, exist_ok=True)
+
+        if private_key_path.exists():
+            private_key_path.unlink()
+        private_key_path.write_bytes(private_key)
+        os.chmod(private_key_path, 0o600)
+
+        public_key_path = Path(self.local_public_key_path).expanduser()
+        if public_key_path.exists():
+            public_key_path.unlink()
+        public_key_path.write_bytes(public_key)
+        os.chmod(public_key_path, 0o600)
+
+    def check_remote_tunnel(self):
+        print("Ping the remote tunnel server...")
+        resp = requests.get(f"{self.tunnel_config.healthcheck_url}",
+                            headers=self._api_client.default_headers)
+        if resp.status_code != 200:
+            raise RuntimeError(f"Unable to connect to the remote tunnel server: {resp.status_code}")
+        status = resp.json()["status"]
+        if status != "working":
+            raise RuntimeError(f"Unexpected remote tunnel server status: {status}")
+
+        return True
+
+    def cleanup(self):
+        self.destroy_context()
+        self._init_params()
+
+    def start_tunneling(self, cluster_id, local_port=None):
```

Comment:
Why are `cluster_id` and `local_port` not provided as arguments for the constructor of `TunnelApi`?

Reply:
Moving `cluster_id` to TunnelApi makes sense. But having `local_port` in `start_tunneling` can be helpful in case we want to reuse TunnelApi to create multiple tunnel connections with different ports